File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/settings/SettingsSwitcher.java
package com.sup.dev.android.views.settings;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.widget.Switch;

import com.sup.dev.android.R;

public class SettingsSwitcher extends SettingsAction {

    private final Switch vSwitcher;

    private OnClickListener onClickListener;
    private boolean salient;

    public SettingsSwitcher(@NonNull Context context) {
        this(context, null);
    }

    public SettingsSwitcher(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        vSwitcher = new Switch(context);
        vSwitcher.setFocusable(false);
        vSwitcher.setOnCheckedChangeListener((v, b) -> {
            setEnabledSubSettings(b);
            if (!salient) onClick();
        });

        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.SettingsSwitcher, 0, 0);
        boolean lineVisible = a.getBoolean(R.styleable.SettingsSwitcher_SettingsSwitcher_lineVisible, true);
        String title = a.getString(R.styleable.SettingsSwitcher_SettingsSwitcher_title);
        String subtitle = a.getString(R.styleable.SettingsSwitcher_SettingsSwitcher_subtitle);
        int icon = a.getResourceId(R.styleable.SettingsSwitcher_SettingsSwitcher_icon, 0);
        boolean checked = a.getBoolean(R.styleable.SettingsSwitcher_SettingsSwitcher_checked, false);
        int iconBackground = a.getColor(R.styleable.SettingsSwitcher_SettingsSwitcher_icon_background, 0);
        a.recycle();

        setLineVisible(lineVisible);
        setTitle(title);
        setSubtitle(subtitle);
        setIcon(icon);
        setChecked(checked);
        setSubView(vSwitcher);
        setIconBackground(iconBackground);

        super.setOnClickListener(v -> {
            setChecked(!vSwitcher.isChecked());
            onClick();
        });
    }

    private void onClick() {
        if (onClickListener != null) onClickListener.onClick(this);
    }

    //
    //  State
    //

    @Override
    public Parcelable onSaveInstanceState() {
        Bundle bundle = new Bundle();
        bundle.putParcelable("SUPER_STATE", super.onSaveInstanceState());
        bundle.putBoolean("checked", isChecked());
        return bundle;
    }

    @Override
    public void onRestoreInstanceState(Parcelable state) {
        if (state instanceof Bundle) {
            Bundle bundle = (Bundle) state;
            salient = true;
            setChecked(bundle.getBoolean("checked"));
            salient = false;
            state = bundle.getParcelable("SUPER_STATE");
        }
        super.onRestoreInstanceState(state);
    }

    //
    //  Setters
    //

    @Override
    public void setOnClickListener(OnClickListener onClickListener) {
        this.onClickListener = onClickListener;
    }


    public void setChecked(boolean checked) {
        salient = true;
        vSwitcher.setChecked(checked);
        salient = false;
        setEnabledSubSettings(checked);
    }

    @Override
    public void setEnabled(boolean enabled) {
        super.setEnabled(enabled);
        vSwitcher.setEnabled(enabled);
    }

    //
    //  Getters
    //

    public boolean isChecked() {
        return vSwitcher.isChecked();
    }


}

File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/settings/SettingsSwitcher.kt
package com.sup.dev.android.views.settings

import android.content.Context
import android.os.Bundle
import android.os.Parcelable
import android.util.AttributeSet
import android.view.View
import android.widget.Switch
import com.sup.dev.android.R
import com.sup.dev.android.tools.ToolsView


class SettingsSwitcher @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : Settings(context, attrs) {

    private val vSwitcher: Switch = Switch(context)

    private var onClickListener: View.OnClickListener? = null
    private var salient: Boolean = false

    //
    //  Getters
    //

    init {

        vSwitcher.isFocusable = false
        vSwitcher.setOnCheckedChangeListener { v, b ->
            setEnabledSubSettings(b)
            if (!salient) onClick()
        }

        val a = context.obtainStyledAttributes(attrs, R.styleable.Settings, 0, 0)
        val checked = a.getBoolean(R.styleable.Settings_Settings_checked, false)
        a.recycle()

        setChecked(checked)
        setSubView(vSwitcher)

        super.setOnClickListener { v ->
            setChecked(!vSwitcher.isChecked)
            onClick()
        }
    }

    private fun onClick() {
        if (onClickListener != null) onClickListener!!.onClick(this)
    }

    //
    //  State
    //

    public fun setChecked(checked:Boolean){
        salient = true
        vSwitcher.isChecked = checked
        salient = false
        setEnabledSubSettings(checked)
    }

    public override fun onSaveInstanceState(): Parcelable? {
        val bundle = Bundle()
        bundle.putParcelable("SUPER_STATE", super.onSaveInstanceState())
        bundle.putBoolean("checked", isChecked())
        return bundle
    }

    public override fun onRestoreInstanceState(state: Parcelable?) {
        var state = state
        if (state is Bundle) {
            val bundle = state as Bundle?
            salient = true
            setChecked(bundle!!.getBoolean("checked"))
            salient = false
            state = bundle.getParcelable("SUPER_STATE")
        }
        super.onRestoreInstanceState(state)
    }

    //
    //  Setters
    //

    override fun setOnClickListener(onClickListener: View.OnClickListener?) {
        this.onClickListener = onClickListener
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)
        vSwitcher.isEnabled = enabled
    }

    //
    //  Getters
    //

    fun isChecked() = vSwitcher.isChecked


}
